AX: Merge AccessibilityARIATable into AccessibilityTable

hoffmanjoshua opened the pull request:
#### 8adbbacf241c8cd19dd73bdefbdb7a7e6b3045bd
<pre>
AX: Merge AccessibilityARIATable into AccessibilityTable
<a href="https://bugs.webkit.org/show_bug.cgi?id=296594">https://bugs.webkit.org/show_bug.cgi?id=296594</a>
<a href="https://rdar.apple.com/156952986">rdar://156952986</a>

Reviewed by Tyler Wilcock.

Merges all of the behavior from ARIA tables into AccessibilityTable.

ARIA tables only had a few behavior differences from regular tables,
so this change allows all of the logic to live in one spot with the
addition of only one new member variable, m_isAriaTable.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIATable.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIATable.h: Removed.
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::AccessibilityTable):
(WebCore::AccessibilityTable::create):
(WebCore::AccessibilityTable::isMultiSelectable const):
* Source/WebCore/accessibility/AccessibilityTable.h:
(WebCore::AccessibilityTable::isAriaTable const):
(WebCore::AccessibilityTable::computeIsTableExposableThroughAccessibility const):

Canonical link: <a href="https://commits.webkit.org/297983@main">https://commits.webkit.org/297983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3e8f40790cb6ebdc94490f67e4ecbb7ab1d7e09

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113738 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33437 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23888 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119900 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64516 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115627 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42004 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86453 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41527 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116686 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102173 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66813 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20302 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63623 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96526 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20456 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123135 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95299 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41121 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98382 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95060 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24259 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40192 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18010 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36939 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40614 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46121 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40252 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43552 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42027 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -848,7 +847,7 @@ Ref<AccessibilityRenderObject> AXObjectCache::createObjectFromRenderer(RenderObj
     // so only create an AccessibilityTable when !is<HTMLTableSectionElement>.
     if ((is<RenderTable>(renderer) && !isAnonymous && !is<HTMLTableSectionElement>(node.get())) || isAccessibilityTable(node.get()))
-        return AccessibilityTable::create(AXID::generate(), renderer, *this);
+        return AccessibilityTable::create(AXID::generate(), renderer, /* isARIATable */ *this);
```
Is this right? The third parameter is not `isARIATable`, it's `AXObjectCache&`, right?

hoffmanjoshua replied:
You're totally right, I was being a little overzealous with that comment 🤦‍♂️

minorninth commented on Source/WebCore/accessibility/AccessibilityTable.cpp:
```diff
@@ -56,17 +56,19 @@ namespace WebCore {
 using namespace HTMLNames;
 
-AccessibilityTable::AccessibilityTable(AXID axID, RenderObject& renderer, AXObjectCache& cache)
+AccessibilityTable::AccessibilityTable(AXID axID, RenderObject& renderer, AXObjectCache& cache, bool isAriaTable)
```
I wonder if we should compute isAriaTable dynamically so that if someone adds role="grid" to a <table> we don't have to recreate the element.

Fine to not do that now. This looks like a simple refactor that preserves the existing behavior.

hoffmanjoshua replied:
That's a good call out. I agree we should do that, but I can file a bug to do that in a separate PR.

hoffmanjoshua replied:
https://bugs.webkit.org/show_bug.cgi?id=296608

minorninth commented on Source/WebCore/accessibility/AccessibilityTable.cpp:
```diff
@@ -410,6 +412,20 @@ void AccessibilityTable::recomputeIsExposable()
 
+bool AccessibilityTable::isMultiSelectable() const
+{
+    if (!m_isAriaTable)
```
This is probably redundant because we have a check for hasGridRole which will be more strict than m_isAriaTable

hoffmanjoshua replied:
Good catch!

hoffmanjoshua committed:
AX: Merge AccessibilityARIATable into AccessibilityTable
https://bugs.webkit.org/show_bug.cgi?id=296594
rdar://156952986

Reviewed by Tyler Wilcock.

Merges all of the behavior from ARIA tables into AccessibilityTable.

ARIA tables only had a few behavior differences from regular tables,
so this change allows all of the logic to live in one spot with the
addition of only one new member variable, m_isAriaTable.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIATable.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIATable.h: Removed.
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::AccessibilityTable):
(WebCore::AccessibilityTable::create):
(WebCore::AccessibilityTable::isMultiSelectable const):
* Source/WebCore/accessibility/AccessibilityTable.h:
(WebCore::AccessibilityTable::isAriaTable const):
(WebCore::AccessibilityTable::computeIsTableExposableThroughAccessibility const):

Canonical link: https://commits.webkit.org/297983@main
